polar-core/src/normalize.rs

```diff
@@ -0,0 +1,317 @@
+
+            Expression(Operation { operator, args }) => match operator {
+                // cancel double negatives
+                Not => self.clone_with_value(args[0].value().clone()),
+
+                // swap and/or using De Morgan's to keep the nots nested
+                // as deep as possible
+                And => {
+                    assert_eq!(args.len(), 2);
+                    let args = vec![args[0].negated(), args[1].negated()];
+                    self.clone_with_value(Expression(Operation { operator: Or, args }))
+                }
+                Or => {
+                    assert_eq!(args.len(), 2);
+                    let args = vec![args[0].negated(), args[1].negated()];
+                    self.clone_with_value(Expression(Operation {
+                        operator: And,
+                        args,
+                    }))
+                }
+                _ => self.clone_with_value(Expression(op!(Not, self.clone()))),
+            },
+            _ => self.clone_with_value(Expression(op!(Not, self.clone()))),
+        }
+    }
+
+    /// binary tree normal form -- all and / or nodes form a binary tree
+    fn btnf(&self) -> Self {
+        use Operator::*;
+        match self.value().as_expression() {
+            Ok(Operation { operator, args }) if *operator == And || *operator == Or => {
+                match args.len() {
+                    // empty -> boolean
+                    0 => self.clone_with_value(Value::Boolean(*operator == And)),
+                    // one -> lift it out
+                    1 => args[0].btnf(),
+                    // else fold
+                    _ => args
+                        .iter()
+                        .map(|a| a.btnf())
+                        .reduce(|m, x| {
+                            self.clone_with_value(Value::Expression(Operation {
+                                operator: *operator,
+                                args: vec![m, x],
+                            }))
+                        })
+                        .unwrap(),
+                }
+            }
+            _ => self.clone(),
+        }
+    }
+
+    /// negation normal form -- all not operators are nested inside of and/or
+    /// using De Morgan's law
+    fn nnf(&self) -> Self {
+        use Operator::*;
+        match self.value().as_expression() {
+            Err(_) => self.clone(),
+            Ok(Operation {
+                operator: Not,
+                args,
+            }) => args[0].nnf().negated(),
+            Ok(Operation { operator, args }) => {
+                self.clone_with_value(Value::Expression(Operation {
+                    operator: *operator,
+                    args: args.iter().cloned().map(|t| t.nnf()).collect(),
+                }))
+            }
+        }
+    }
+
+    fn _hs(&self, n: usize) -> Self {
+        self.value().as_expression().unwrap().args[n].clone()
+    }
+
+    fn lhs(&self) -> Self {
```

Comment:
assert self len is 2? Otherwise LHS & RHS doesn't have meaning.

Reply:
Already assumed because of the prior call to `btnf`

Reply:
I'm going to nest the function defs so it's impossible to call them in the wrong way 👍

Reply:
nice idea